temporal-sdk/src/main/java/io/temporal/internal/replay/ReplayWorkflowExecutor.java

```
@@ -153,7 +154,9 @@ private void completeWorkflow(@Nullable WorkflowExecutionException failure) {
       metricsScope.counter(MetricsType.WORKFLOW_CANCELED_COUNTER).inc(1);
     } else if (failure != null) {
       workflowStateMachines.failWorkflow(failure.getFailure());
-      metricsScope.counter(MetricsType.WORKFLOW_FAILED_COUNTER).inc(1);
+      if (!ApplicationFailure.isBenignApplicationFailure(failure)) {
```

I do not think you want to swallow activities that fail because of benign exceptions. They may reach a max attempts or something and throw this out. Just because an activity has a benign exception it doesn't want to be in telemetry doesn't mean a workflow doesn't want it in telemetry.

---

Is this not the same pattern in:
https://github.com/temporalio/sdk-go/pull/1925
https://github.com/temporalio/sdk-core/pull/905
unless I'm mistaken.

What about if a user throws a benign exception in workflow code, not in an activity?

---

Sorry, I was unclear here. This is basically me restating https://github.com/temporalio/sdk-java/pull/2485/files#r2054385164.

> What about if a user throws a benign exception in workflow code, not in an activity?

That is fine to skip metrics, but we don't want to skip metrics if the exception _in the workflow code_ is an activity failure with a benign cause. There is a difference between not recording telemetry on benign exceptions and not recording telemetry on an exception that has a benign cause.

Here are the scenarios:

* Activity throws benign exception - do not record telemetry
* Workflow throws benign exception - do not record telemetry
* Activity or child fails with benign exception from workflow POV (thrown from activity and it is out of max attempts or something) - still record telemetry

Basically, change `isBenignApplicationFailure` to stop checking causes IMO. IMO throwing a benign exception is benign, but receiving a benign exception wrapped in another exception is not benign.

> Is this not the same pattern in:
> https://github.com/temporalio/sdk-go/pull/1925
> https://github.com/temporalio/sdk-core/pull/905
> Unless I'm mistaken.

Not exactly because Go stops at the first application error, not every application error no matter the depth. Also, this isn't recursively checking causes no matter how deep. Still, looking back on Go, we should only apply it to the current error and not check cause or `errors.As`. Didn't check Go, but I think the same applies there too.

If you think about how users use benign exceptions (as control flow), they aren't supposed to wrap them.

---

Won't they inherently be wrapped as other failure types though? Could be wrong here, but for example, an ApplicationFailure thrown from an activity, would it not always (or at least in the general case) be wrapped as an ActivityFailure?

---

I agree in that it doesn't seem relevant/useful to check for benign causes at an arbitrary depth, but I though a depth of 0 (i.e. immediate) or 1 would be sufficient.

In Java SDK, how would one receive an ApplicationFailure that is **not** wrapped?

---

> Won't they inherently be wrapped as other failure types though?

No. They are only benign where they are thrown (impl side) and they are usually only wrapped where they are caught (caller side).

> Could be wrong here, but for example, an ApplicationFailure thrown from an activity, would it not always (or at least in the general case) be wrapped as an ActivityFailure?

Not on the activity side, only on the workflow caller side. But if an activity fails because of a benign error, even though it is benign on the activity side, if that activity failure bubbles out of a workflow, that is _not_ benign.

> In Java SDK, how would one receive an ApplicationFailure that is not wrapped?

What do you mean by "one"? If you mean how does our SDK internals receive an unwrapped thrown error, it is where you have changes in `ActivityTaskExecutors.java`. If you mean how does a workflow caller receive it unwrapped, they don't, that's the point.

An error is only benign where it is thrown impl side, not where it is caught caller side.

---

Thanks for the clarification, that all sounds reasonable (and saved me some digging :))
I've removed the nested failure checks, only check immediate failures.